### ISSUE TITLE
Server crashes with specific GET request

### ISSUE THREAD

**Issue by serjek** [haxetink/tink_web#107]
Following server setup:
```haxe
var container:NodeContainer = new NodeContainer(3567);
var router:Router<Root> = new Router<Root>(new Root());

container.run(req -> {
	router
		.route(Context.ofRequest(req))
		.recover(OutgoingResponse.reportError);
	}
);

```
In browser navigate to:
http://0.0.0.0:3567//%80../%80../%80../%80../%80../%80../windows/win.ini

server will crash with following error:
```
 return decodeURIComponent(this1.split("+").join(" "));
                       ^
URIError: URI malformed
    at decodeURIComponent (<anonymous>)
    at Object.tink_url__$Portion_Portion_$Impl_$.toString 
```

**Comment by serjek** [comment]
wrapping `router.route` into `try...catch` solves the problem:
```haxe
try
	router
		.route(Context.ofRequest(req))
		.recover(OutgoingResponse.reportError)
catch (e:Dynamic) {
	Future.sync(OutgoingResponse.reportError(new Error(Std.string(e)));
} 
```

**Comment by back2dos** [comment]
This workaround only catches the error if routing is fully synchronous.

We could:

1. Catch the error in `tink.url.Portion.toString` (return original if urlDecode fails), which will lead to a 404 in routing or pass gibberish to a path parameter.
2. Catch the error during routing and report a 422. (preferable)

**Comment by emugel** [comment]
Regarding above bug, I wonder if there is something that runs **before** middlewares that could prematurely reject uris containing null bytes e.g. with a 422. 

Because: 

1. they suck (in C language, string end with null bytes, creating security threats). 
2. Also there is no good way to fix them, if code A removes them and code B keeps them, it creates inconsistencies and some tests may in turn creating more security threats.
3. Middlewares are chained and are optional anyway, so it seems all pointless to do something at this level. And to make matter worse people can write their own middleware. 

In short, is it possible to immediately e.g. `if s.indexOf("\x00") > -1` somewhere then `422`?

**Comment by emugel** [comment]
Appears web servers actually never reject uris with params containing null bytes. In node you can even have a route containing null bytes e.g. `/%00hi%00`. 

Therefore it is the problem of middlewares. Sorry if I disturbed :)

**Comment by back2dos** [comment]
Null bytes are "fine" I guess. The thing is percent encoding itself is only valid upto 0x7F.

**Comment by emugel** [comment]
Hum. Yes I agree `%80` and above don't have the excuses for himself that `%00` has. 
A rejection seems ok in that case. Though it is always tough to answer those questions in a hurry.

**Comment by emugel** [comment]
Actually: 
* `GET /onehundred%`
* `GET /%WTF`

both crash as well as the above `%80`.

**Comment by emugel** [comment]
Just modified  `SimpleHandler` in `tink.http.Handler` like so:

```haxe
public function process(req:IncomingRequest):Future<OutgoingResponse>
    return StringTools.urlDecode.bind(req.header.url).catchExceptions().isSuccess()
        ? f(req)
        : OutgoingResponse.reportError(new Error(UnprocessableEntity, "Unprocessable Entity"))
    ;
```
The above fixes the malformed URI problem. But I am not sure if this is an acceptable solution, so I just propose this for discussion. Anyway for now the 7688 attacks from the nikto suite don't make my server crash anymore.

**Comment by back2dos** [comment]
Ok, so using tink_url >= 0.4.2 this won't throw anymore (invalid portions are decoded as empty strings). If anyone feels strongly in favor of producing a 422, PRs are welcome ;)

**Comment by emugel** [comment]
Before this gets cold and we both forget, 
by this
| If anyone feels strongly in favor of producing a 422, PRs are welcome ;)
were you talking about my snippet just above or another solution?  :)

**Comment by back2dos** [comment]
> were you talking about my snippet just above or another solution?

From my side, I think it's fine. Not exactly glorious, but it gets the job done.

**Comment by emugel** [comment]
Yes I think just the same. 
I will perform more tests soon, e.g. with unicode uris. So will see if the same recipe has to be applied again. For now I think is is not urgently needed.
